cmd/kubelet/app/server.go

```diff
@@ -186,6 +186,120 @@ func Run(s *options.KubeletServer, kubeDeps *kubelet.Dependencies) error {
 	return nil
 }
 
+// setKubeDepsClients initializes clients if not standalone mode and any of the clients are not provided
+func setKubeDepsClients(s *options.KubeletServer, kubeDeps *kubelet.Dependencies, nodeName types.NodeName) error {
+	clientConfig, err := CreateAPIServerClientConfig(s)
+	if err == nil {
+		clientCertificateManager, err := makeKubeletClientCertificateManager(s, clientConfig, nodeName)
+		if err != nil {
+			return err
+		}
+
+		kubeClient, err := makeKubeDepsKubeClient(kubeDeps, clientConfig)
+		if err != nil {
+			startClientCertificateManager(kubeClient, clientCertificateManager)
+		}
+
+		setKubeDepsExternalKubeClient(kubeDeps, clientConfig)
+		setKubeDepsEventClient(s, kubeDeps, clientConfig)
+		setKubeDepsHeartbeatClient(s, kubeDeps, clientConfig)
+
+	} else {
+		switch {
+		case s.RequireKubeConfig:
+			return fmt.Errorf("invalid kubeconfig: %v", err)
+		case s.KubeConfig.Provided():
+			glog.Warningf("invalid kubeconfig: %v", err)
+		}
+	}
+
+	return nil
+}
+
+func makeKubeletClientCertificateManager(
+	s *options.KubeletServer,
```

Comment:
Rather than passing in the whole kubelet server object, please just pass in the parameters you need. In this case it looks like all you need is `s.CertDirectory`.

Reply:
And rotateCertificates  :)